Add config option to change error.log location (+ optionally better XDG support)

ayekat opened the pull request:
These commit add the `log_directory` option to the ncmpcpp configuration, which allows the user to set the location of `error.log` independently from `ncmpcpp_directory`.

If `log_directory` is not specified, `error.log` will be stored into `$XDG_DATA_HOME/ncmpcpp` (note that unfortunately, the [XDG base directory specification][1] does not mention log files at all).

This commit is just a proposal. I would like to have some feedback on which default value for `log_directory` makes most sense:

1. `$XDG_DATA_HOME/ncmpcpp/error.log` [current]: Treat logs like regular data/state files (see issue #110).
2. `$XDG_CACHE_HOME/ncmpcpp/error.log`: Treat logs like cache files (this is semantically incorrect IMHO, but there are people who might still argue for this).
2. `$XDG_LOG_HOME/ncmpcpp/error.log`: Note that `$XDG_LOG_HOME` is non-standard. However, some people seem to set and use it (even though [not very uniformly][2]). We would certainly need to agree on a default value for this.

**P.S.**
~~With this change, it appears that `ncmpcpp_directory` is no longer used for anything, also since the location for the `bindings` file respects `$XDG_CONFIG_HOME` now ([pull request 165][3]). Is it safe to remove `ncmpcpp_directory`?~~
Nevermind, I didn't see the tag editor.

[1]: https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html
[2]: https://encrypted.google.com/search?hl=en&q=%22XDG_LOG_HOME%22#q=%22XDG_LOG_HOME%22&hl=en&filter=0
[3]: https://github.com/arybczak/ncmpcpp/pull/165

olmokramer commented:
I think it makes more sense to use the value of `ncmpcpp_directory` as the default. It's compatible with current behavior and doesn't break expectations for current users. Other than that this would be nice to have :)

ayekat commented:
I went for that initially, but it turns out that by the moment I define the `log_directory` configuration variable (and a default path), the value for `ncmpcpp_directory` is not known yet, so I can't fall back to it for the default path yet. But I'm not 100% sure—perhaps it is possible somehow; I just don't know how.

olmokramer commented:
Ah ok, I see, I'm not that familiar with the code, so I thought I'd mention it :) But this made me think: wouldn't it be better if I could set the exact *file* path instead of the *directory*, i.e. `$XDG_CACHE_HOME/ncmpcpp.log` instead of just `$XDG_CACHE_HOME`?

ayekat commented:
The other configuration options (`ncmpcpp_directory`, `lyrics_directory`) also specify directories, so I thought this would be preferred. But I don't know what @arybczak thinks about this.

ayekat commented:
I have updated the commits to store the logs into `XDG_DATA_HOME` by default (rather than `XDG_CACHE_HOME`, as before), as I believe that it would be semantically incorrect to have files that cannot be easily regenerated (state files, logs, etc.) stored as "cache".

I have also split the changes for getting an environment variable into a separate commit, as it is independent from this "where do we put the log files" matter.

If you prefer not changing the default location, feel free to only merge the first commit (which simply adds a configuration option to change the location). The second and third commit would just be the cherry-on-top for all the people who would prefer applications following the XDG basedir spec.

larson-carter commented:
@ayekat This seems very useful. I will need to do some more extensive testing tomorrow when I have a little more time.

larson-carter commented:
I'm going to close this for right now. Once Author becomes more active I can merge into the branch.

ayekat commented:
Rebased onto the current master.

ayekat commented:
Rebased onto the current master again. Are there plans to consider this for merging, or is this feature not desired?

larson-carter commented:
Please reopen this PR against our current branch.

ayekat commented:
It is already based on the current master. This pull request can simply be reopened. Or what am I missing?

ayekat commented:
Maybe GitHub doesn't allow reopening a closed pull request, so I've created #359.

thepenguinthatwants commented:
Is this going to be enabled soon?
Tried
` log_directory = ~/.local/share/ncmpcpp`
but it doesnt seem to work.

ayekat commented:
@thepenguinthatwants  No, the `log_directory` idea was discarded. See #359 for a continuation of this (this PR was closed, not merged).
